File: bindings/nodejs/examples/how_tos/nft_collection/01_mint_collection_nft.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { MintNftParams, NftId, Utils, Wallet, Irc27Metadata } from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// The NFT collection size
const NFT_COLLECTION_SIZE = 150;
// Mint NFTs in chunks since the transaction size is limited
const NUM_NFTS_MINTED_PER_TRANSACTION = 50;

// In this example we will mint the issuer NFT for the NFT collection.
//
// Make sure that `STRONGHOLD_SNAPSHOT_PATH` and `WALLET_DB_PATH` already exist by
// running the `how_tos/accounts_and_addresses/create-account` example!
//
// Rename `.env.example` to `.env` first, then run
// yarn run-example ./how_tos/nfts/01_mint_collection_nft.ts <issuer_nft_id>
async function run() {
    try {
        for (const envVar of [
            'WALLET_DB_PATH',
            'STRONGHOLD_PASSWORD',
            'EXPLORER_URL',
        ]) {
            if (!(envVar in process.env)) {
                throw new Error(
                    `.env ${envVar} is undefined, see .env.example`,
                );
            }
        }

        // Create the wallet
        const wallet = new Wallet({
            storagePath: process.env.WALLET_DB_PATH,
        });

        // To sign a transaction we need to unlock stronghold.
        await wallet.setStrongholdPassword(
            process.env.STRONGHOLD_PASSWORD as string,
        );

        // Get the account we generated with `how_tos/accounts_and_addresses/create-account`
        const account = await wallet.getAccount('Alice');

        await account.sync();
        console.log(`Account synced!`);

        // Get the id we generated with `00_mint_issuer_nft`
        const issuerNftId: NftId = process.argv[2];

        const bech32Hrp = await (await wallet.getClient()).getBech32Hrp();
        const issuer = Utils.nftIdToBech32(issuerNftId, bech32Hrp);

        const nftMintParams = [];
        // Create the metadata with another index for each
        for (let index = 0; index < NFT_COLLECTION_SIZE; index++) {
            const params: MintNftParams = {
                immutableMetadata: getImmutableMetadata(index).asHex(),
                // The NFT address from the NFT we minted in mint_issuer_nft example
                issuer,
            };
            nftMintParams.push(params);
        }

        for (
            let i = 0;
            i < nftMintParams.length;
            i += NUM_NFTS_MINTED_PER_TRANSACTION
        ) {
            const chunk = nftMintParams.slice(
                i,
                i + NUM_NFTS_MINTED_PER_TRANSACTION,
            );

            console.log(
                `Minting ${chunk.length} NFTs... (${
                    i + chunk.length
                }/${NFT_COLLECTION_SIZE})`,
            );
            const transaction = await account.mintNfts(chunk);

            // Wait for transaction to get included
            const blockId = await account.retryTransactionUntilIncluded(
                transaction.transactionId,
            );
            console.log(
                `Block included: ${process.env.EXPLORER_URL}/block/${blockId}`,
            );

            // Sync so the new outputs are available again for new transactions
            await account.sync();
        }

        // After the NFTs are minted, the issuer nft can be sent to the so called "null address"
        // 0x0000000000000000000000000000000000000000000000000000000000000000 (for smr:
        // smr1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqy8f002) or burned, to
        // prevent minting any further NFTs in the future. Sending it to the null address makes it still available to get
        // its metadata.
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

function getImmutableMetadata(index: number) {
    return new Irc27Metadata(
        'video/mp4',
        'https://ipfs.io/ipfs/QmPoYcVm9fx47YXNTkhpMEYSxCD3Bqh7PJYr7eo5YjLgiT',
        `Shimmer OG NFT ${index}`,
    )
        .withDescription(
            'The Shimmer OG NFT was handed out 1337 times by the IOTA Foundation \
        to celebrate the official launch of the Shimmer Network.',
        )
        .withIssuerName('IOTA Foundation')
        .withCollectionName('Shimmer OG');
}

run();

File: bindings/python/examples/how_tos/nft_collection/01_mint_collection_nft.py
import os
import sys

from dotenv import load_dotenv
from iota_sdk import MintNftParams, Utils, Wallet, WalletOptions, Irc27Metadata

load_dotenv()

# The NFT collection size
NFT_COLLECTION_SIZE = 150
# Mint NFTs in chunks since the transaction size is limited
NUM_NFTS_MINTED_PER_TRANSACTION = 50

# In this example we will mint some collection NFTs with issuer feature.

if len(sys.argv) < 2:
    raise Exception("missing example argument: ISSUER_NFT_ID")

issuer_nft_id = sys.argv[1]

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception(".env STRONGHOLD_PASSWORD is undefined, see .env.example")

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

# Sync wallet with the node
wallet.sync()

bech32_hrp = wallet.get_client().get_bech32_hrp()
issuer = Utils.nft_id_to_bech32(issuer_nft_id, bech32_hrp)


def get_immutable_metadata(index: int) -> str:
    """Returns the immutable metadata for the NFT with the given index"""
    Irc27Metadata(
        "video/mp4",
        "https://ipfs.io/ipfs/QmPoYcVm9fx47YXNTkhpMEYSxCD3Bqh7PJYr7eo5YjLgiT",
        "Shimmer OG NFT #" + str(index),
        description="The Shimmer OG NFT was handed out 1337 times by the IOTA Foundation to celebrate the official launch of the Shimmer Network.",
        issuerName="IOTA Foundation",
        collectionName="Shimmer OG",
    ).as_hex()


# Create the metadata with another index for each
nft_mint_params = list(
    map(
        lambda index: MintNftParams(
            immutable_metadata=get_immutable_metadata(index), issuer=issuer
        ),
        range(NFT_COLLECTION_SIZE),
    )
)

while nft_mint_params:
    chunk, nft_mint_params = (
        nft_mint_params[:NUM_NFTS_MINTED_PER_TRANSACTION],
        nft_mint_params[NUM_NFTS_MINTED_PER_TRANSACTION:],
    )
    print(
        f'Minting {len(chunk)} NFTs... ({NFT_COLLECTION_SIZE-len(nft_mint_params)}/{NFT_COLLECTION_SIZE})'
    )
    transaction = wallet.mint_nfts(chunk)

    wallet.wait_for_transaction_acceptance(
        transaction.transaction_id)

    print(
        f'Tx accepted: {os.environ["EXPLORER_URL"]}/transactions/{transaction.transaction_id}')

    # Sync so the new outputs are available again for new transactions
    wallet.sync()
